File: Nbody/Analysis/plot_virial_evolution_from_db.py
#!/usr/bin/env python3
import sys
import argparse
import numpy as np
import matplotlib.pyplot as plt
import psycopg2

def get_bound_subsystem(pos, vel, mass, G, softening, max_iter=10):
    """
    Iteratively removes unbound stars (E_i = T_i + V_i >= 0) to find the bound subsystem,
    and returns its indices, total kinetic energy, total potential energy, and bound count.
    """
    idx_bound = np.arange(len(mass))
    
    T_tot = 0.0
    V_tot = 0.0
    
    for iteration in range(max_iter):
        N_curr = len(idx_bound)
        if N_curr < 3:
            return idx_bound, 0.0, 0.0
            
        pos_curr = pos[idx_bound]
        vel_curr = vel[idx_bound]
        mass_curr = mass[idx_bound]
        
        # Center of mass position & velocity of current bound subset
        total_m = np.sum(mass_curr)
        cm_pos = np.sum(pos_curr * mass_curr[:, np.newaxis], axis=0) / total_m
        cm_vel = np.sum(vel_curr * mass_curr[:, np.newaxis], axis=0) / total_m
        
        # Relative coordinates
        rel_pos = pos_curr - cm_pos
        rel_vel = vel_curr - cm_vel
        
        # Kinetic energy of each star relative to bound CM
        T_i = 0.5 * mass_curr * np.sum(rel_vel**2, axis=1)
        
        # Potential of each star due to other bound stars
        pos_sq = np.sum(rel_pos**2, axis=1)
        dist_sq = pos_sq[:, np.newaxis] + pos_sq[np.newaxis, :] - 2 * np.dot(rel_pos, rel_pos.T)
        dist_sq = np.maximum(dist_sq, 0.0)  # Avoid negative values due to floating-point precision
        inv_dist = 1.0 / np.sqrt(dist_sq + softening**2)
        np.fill_diagonal(inv_dist, 0)
        
        Phi_i = -G * np.sum(mass_curr * inv_dist, axis=1)
        V_i = mass_curr * Phi_i
        
        # Total energy of each star
        E_i = T_i + V_i
        
        # Identify bound stars
        bound_mask = E_i < 0
        N_next = np.sum(bound_mask)
        
        T_tot = np.sum(T_i)
        V_tot = 0.5 * np.sum(V_i) # 0.5 to avoid double counting pairs
        
        if N_next == N_curr:
            break
            
        idx_bound = idx_bound[bound_mask]
        
    return idx_bound, T_tot, V_tot

def main():
    parser = argparse.ArgumentParser(description="Plot the evolution of the virial ratio of bound stars from the hypercluster database.")
    parser.add_argument("--dim", type=int, help="Spatial dimension of the simulation (e.g., 2, 3, 4, 5, 6, 100)")
    parser.add_argument("--host", type=str, default="localhost", help="Database host")
    parser.add_argument("--user", type=str, default="stephane", help="Database user")
    parser.add_argument("--password", type=str, default="tallis", help="Database password")
    parser.add_argument("--dbname", type=str, default="hypercluster", help="Database name")
    parser.add_argument("--softening", type=float, default=0.01, help="Softening parameter used in simulation (default: 0.01)")
    parser.add_argument("--g-constant", type=float, default=1.0, help="Gravitational constant G (default: 1.0)")
    parser.add_argument("--max-points", type=int, default=200, help="Maximum number of snapshots to analyze to keep runtimes fast (default: 200)")
    args = parser.parse_args()

    # Prompt if dimension not provided
    dim_space = args.dim
    if dim_space is None:
        try:
            dim_space = int(input("Enter the spatial dimension (dim_space) to plot (e.g., 2, 3, 4, 5, 6, 100): "))
        except ValueError:
            print("Invalid input. Please enter an integer.")
            sys.exit(1)

    print(f"Connecting to database '{args.dbname}' on '{args.host}'...")
    try:
        conn = psycopg2.connect(
            dbname=args.dbname,
            user=args.user,
            password=args.password,
            host=args.host
        )
        cursor = conn.cursor()
    except Exception as e:
        print(f"Failed to connect to database: {e}")
        sys.exit(1)

    print(f"Connected successfully. Querying run history for dim_space = {dim_space}...")

    # 1. Find the maximum snapshot_id for this dimension (belongs to the latest run)
    cursor.execute("""
        SELECT MAX(snapshot_id) 
        FROM star_snapshots 
        WHERE dim_space = %s;
    """, (dim_space,))
    max_snapshot_id = cursor.fetchone()[0]
    
    if max_snapshot_id is None:
        print(f"No simulation data found in the database for dim_space = {dim_space}.")
        cursor.close()
        conn.close()
        sys.exit(1)

    # 2. Get the number of stars from the latest snapshot
    cursor.execute("""
        SELECT COUNT(*) 
        FROM star_snapshots 
        WHERE dim_space = %s AND snapshot_id = %s;
    """, (dim_space, max_snapshot_id))
    num_stars = cursor.fetchone()[0]

    # Find the minimum snapshot_id for this dimension
    cursor.execute("""
        SELECT MIN(snapshot_id)
        FROM star_snapshots
        WHERE dim_space = %s;
    """, (dim_space,))
    min_snapshot_id = cursor.fetchone()[0]

    # 3. Find the starting database ID for this run (minimum id of the latest min_snapshot_id)
    cursor.execute("""
        SELECT MIN(id) FROM (
            SELECT id FROM star_snapshots 
            WHERE dim_space = %s AND snapshot_id = %s 
            ORDER BY id DESC 
            LIMIT %s
        ) as sub;
    """, (dim_space, min_snapshot_id, num_stars))
    start_db_id = cursor.fetchone()[0]

    # 4. Find the step size dynamically
    cursor.execute("""
        SELECT MIN(snapshot_id) 
        FROM star_snapshots 
        WHERE dim_space = %s AND snapshot_id > %s AND id >= %s;
    """, (dim_space, min_snapshot_id, start_db_id))
    second_snap = cursor.fetchone()[0]
    step_size = (second_snap - min_snapshot_id) if second_snap is not None else 100

    # Generate full snapshot IDs list
    snap_ids = list(range(min_snapshot_id, max_snapshot_id + 1, step_size))
    total_snapshots = len(snap_ids)

    print(f"Detected snapshots up to ID {max_snapshot_id} (step size: {step_size}). Total potential snapshots: {total_snapshots}")

    # Downsample snapshots
    sample_interval = max(1, total_snapshots // args.max_points)
    sampled_snap_ids = snap_ids[::sample_interval]
    print(f"Downsampled to {len(sampled_snap_ids)} snapshots for analysis (interval: every {sample_interval} steps).")

    times = []
    virial_ratios = []
    virial_ratios_all = []
    bound_fractions = []

    # Get N_STARS
    cursor.execute("""
        SELECT COUNT(*) 
        FROM star_snapshots 
        WHERE dim_space = %s AND snapshot_id = %s AND id >= %s;
    """, (dim_space, sampled_snap_ids[0], start_db_id))
    n_stars = cursor.fetchone()[0]
    print(f"Number of stars per snapshot: {n_stars}")

    # Process each snapshot
    for idx, snap_id in enumerate(sampled_snap_ids):
        sys.stdout.write(f"\rAnalyzing snapshot {idx+1}/{len(sampled_snap_ids)} (ID: {snap_id})...")
        sys.stdout.flush()
        
        # Load positions, velocities, masses, and time for this snapshot
        cursor.execute("""
            SELECT position, velocity, mass, time_myr FROM (
                SELECT id, position, velocity, mass, time_myr 
                FROM star_snapshots 
                WHERE dim_space = %s AND snapshot_id = %s AND id >= %s
                ORDER BY id DESC 
                LIMIT %s
            ) as sub 
            ORDER BY id ASC;
        """, (dim_space, snap_id, start_db_id, n_stars))
        
        rows = cursor.fetchall()
        if len(rows) == 0:
            continue
            
        pos = np.array([row[0] for row in rows])
        vel = np.array([row[1] for row in rows])
        mass = np.array([row[2] for row in rows])
        # Normalize masses to N-body units (sum to 1.0) to match N-body positions and velocities
        mass = mass / np.sum(mass)
        time_val = float(rows[0][3])
        
        # 1. Compute bound subsystem virial
        bound_indices, T_bound, V_bound = get_bound_subsystem(
            pos, vel, mass, args.g_constant, args.softening
        )
        if V_bound != 0:
            Q_bound = T_bound / abs(V_bound)
        else:
            Q_bound = np.nan
            
        # 2. Compute entire cluster virial (all stars)
        total_m_all = np.sum(mass)
        cm_pos_all = np.sum(pos * mass[:, np.newaxis], axis=0) / total_m_all
        cm_vel_all = np.sum(vel * mass[:, np.newaxis], axis=0) / total_m_all
        rel_pos_all = pos - cm_pos_all
        rel_vel_all = vel - cm_vel_all
        
        T_all = 0.5 * np.sum(mass * np.sum(rel_vel_all**2, axis=1))
        
        pos_sq_all = np.sum(rel_pos_all**2, axis=1)
        dist_sq_all = pos_sq_all[:, np.newaxis] + pos_sq_all[np.newaxis, :] - 2 * np.dot(rel_pos_all, rel_pos_all.T)
        dist_sq_all = np.maximum(dist_sq_all, 0.0)
        inv_dist_all = 1.0 / np.sqrt(dist_sq_all + args.softening**2)
        np.fill_diagonal(inv_dist_all, 0)
        
        Phi_all = -args.g_constant * np.sum(mass * inv_dist_all, axis=1)
        V_all = 0.5 * np.sum(mass * Phi_all)
        
        if V_all != 0:
            Q_all = T_all / abs(V_all)
        else:
            Q_all = np.nan
            
        bound_frac = len(bound_indices) / len(mass)
        
        times.append(time_val)
        virial_ratios.append(Q_bound)
        virial_ratios_all.append(Q_all)
        bound_fractions.append(bound_frac)

    print("\nAnalysis complete. Generating plots...")
    cursor.close()
    conn.close()

    # Create figure with 2 subplots (Virial Ratio and Bound Fraction)
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 8), sharex=True)

    # Plot Virial Ratio
    ax1.plot(times, virial_ratios, color='crimson', label='Bound Core Virial Q = T/|V|', linewidth=2)
    ax1.plot(times, virial_ratios_all, color='darkorange', linestyle='--', label='Entire Cluster Virial Q = T/|V|', linewidth=2)
    ax1.axhline(y=0.5, color='gray', linestyle=':', label='Virial Equilibrium (Q=0.5)')
    ax1.set_ylabel('Virial Ratio Q')
    ax1.set_title(f'Evolution of the Virial Ratio (N={dim_space} dimensions)')
    ax1.legend()
    ax1.grid(True, linestyle='--', alpha=0.5)

    # Plot Bound Fraction
    ax2.plot(times, np.array(bound_fractions) * 100, color='royalblue', label='Bound Star %', linewidth=2)
    ax2.set_xlabel('Time (Myr)')
    ax2.set_ylabel('Bound Stars (%)')
    ax2.legend()
    ax2.grid(True, linestyle='--', alpha=0.5)

    plt.tight_layout()
    plot_filename = f'virial_evolution_{dim_space}D.png'
    plt.savefig(plot_filename, dpi=300, bbox_inches='tight')
    print(f"Plot successfully saved to '{plot_filename}'!")

if __name__ == "__main__":
    main()

File: Nbody/Analysis/plot_virial_evolution_from_db.jl
#!/usr/bin/env julia

import Pkg
try
    using LibPQ, DataFrames, ArgParse, Plots
catch
    println("Installing missing dependencies...")
    Pkg.add(["LibPQ", "DataFrames", "ArgParse", "Plots"])
    using LibPQ, DataFrames, ArgParse, Plots
end

function get_bound_subsystem(pos::Matrix{Float64}, vel::Matrix{Float64}, mass::Vector{Float64}, G::Float64, softening::Float64; max_iter::Int=10)
    idx_bound = collect(1:length(mass))
    
    T_tot = 0.0
    V_tot = 0.0
    
    for iteration in 1:max_iter
        N_curr = length(idx_bound)
        if N_curr < 3
            return idx_bound, 0.0, 0.0
        end
        
        pos_curr = pos[idx_bound, :]
        vel_curr = vel[idx_bound, :]
        mass_curr = mass[idx_bound]
        
        total_m = sum(mass_curr)
        cm_pos = sum(pos_curr .* mass_curr, dims=1) ./ total_m
        cm_vel = sum(vel_curr .* mass_curr, dims=1) ./ total_m
        
        rel_pos = pos_curr .- cm_pos
        rel_vel = vel_curr .- cm_vel
        
        # Kinetic energy of each star
        T_i = 0.5 .* mass_curr .* vec(sum(rel_vel.^2, dims=2))
        
        # Potential energy of each star
        V_i = zeros(N_curr)
        
        Threads.@threads for i in 1:N_curr
            phi_i = 0.0
            for j in 1:N_curr
                if i != j
                    dist_sq = 0.0
                    for d in 1:size(pos_curr, 2)
                        dist_sq += (rel_pos[i, d] - rel_pos[j, d])^2
                    end
                    phi_i += mass_curr[j] / sqrt(dist_sq + softening^2)
                end
            end
            V_i[i] = -G * mass_curr[i] * phi_i
        end
        
        E_i = T_i .+ V_i
        
        bound_mask = E_i .< 0
        N_next = sum(bound_mask)
        
        T_tot = sum(T_i)
        V_tot = 0.5 * sum(V_i)
        
        if N_next == N_curr
            break
        end
        
        idx_bound = idx_bound[bound_mask]
    end
    
    return idx_bound, T_tot, V_tot
end

function parse_commandline()
    s = ArgParseSettings(description="Plot the evolution of the virial ratio of bound stars from the hypercluster database (Julia version).")

    @add_arg_table! s begin
        "--dim"
            help = "Spatial dimension of the simulation (e.g., 2, 3, 4, 5, 6, 100)"
            arg_type = Int
        "--host"
            help = "Database host"
            default = "localhost"
        "--user"
            help = "Database user"
            default = "stephane"
        "--password"
            help = "Database password"
            default = "tallis"
        "--dbname"
            help = "Database name"
            default = "hypercluster"
        "--softening"
            help = "Softening parameter used in simulation (default: 0.001)"
            arg_type = Float64
            default = 0.001
        "--g-constant"
            help = "Gravitational constant G (default: 1.0)"
            arg_type = Float64
            default = 1.0
        "--max-points"
            help = "Maximum number of snapshots to analyze (default: 200)"
            arg_type = Int
            default = 200
    end

    return parse_args(s)
end

function main()
    args = parse_commandline()
    
    dim_space = args["dim"]
    if dim_space === nothing
        print("Enter the spatial dimension (dim_space) to plot (e.g., 2, 3, 4, 5, 6, 100): ")
        dim_space_str = readline()
        dim_space = parse(Int, dim_space_str)
    end
    
    println("Connecting to database '$(args["dbname"])' on '$(args["host"])'...")
    
    conn_str = "dbname=$(args["dbname"]) user=$(args["user"]) password=$(args["password"]) host=$(args["host"])"
    conn = try
        LibPQ.Connection(conn_str)
    catch e
        println("Failed to connect to database: ", e)
        exit(1)
    end
    
    println("Connected successfully. Querying run history for dim_space = $dim_space...")
    
    # 1. max snapshot
    result = execute(conn, "SELECT MAX(snapshot_id) FROM star_snapshots WHERE dim_space = $dim_space;")
    max_snapshot_id = first(result).max
    if ismissing(max_snapshot_id)
        println("No simulation data found in the database for dim_space = $dim_space.")
        close(conn)
        exit(1)
    end
    
    # 2. min snapshot
    result = execute(conn, "SELECT MIN(snapshot_id) FROM star_snapshots WHERE dim_space = $dim_space;")
    min_snapshot_id = first(result).min
    
    # 3. get number of stars from the final snapshot to verify
    result = execute(conn, "SELECT COUNT(*) FROM star_snapshots WHERE dim_space = $dim_space AND snapshot_id = $max_snapshot_id;")
    num_stars = first(result).count
    
    # 4. start db id
    query_start_id = """
        SELECT MIN(id) FROM (
            SELECT id FROM star_snapshots 
            WHERE dim_space = $dim_space AND snapshot_id = $min_snapshot_id 
            ORDER BY id DESC 
            LIMIT $num_stars
        ) as sub;
    """
    result = execute(conn, query_start_id)
    start_db_id = first(result).min
    
    # 5. step size
    query_step = """
        SELECT MIN(snapshot_id) 
        FROM star_snapshots 
        WHERE dim_space = $dim_space AND snapshot_id > $min_snapshot_id AND id >= $start_db_id;
    """
    result = execute(conn, query_step)
    second_snap = first(result).min
    step_size = !ismissing(second_snap) ? (second_snap - min_snapshot_id) : 100
    
    snap_ids = collect(min_snapshot_id:step_size:max_snapshot_id)
    total_snapshots = length(snap_ids)
    
    println("Detected snapshots up to ID $max_snapshot_id (step size: $step_size). Total potential snapshots: $total_snapshots")
    
    sample_interval = max(1, total_snapshots ÷ args["max-points"])
    sampled_snap_ids = snap_ids[1:sample_interval:end]
    println("Downsampled to $(length(sampled_snap_ids)) snapshots for analysis (interval: every $sample_interval steps).")
    
    # get n_stars for actual query limit
    query_nstars = """
        SELECT COUNT(*) 
        FROM star_snapshots 
        WHERE dim_space = $dim_space AND snapshot_id = $(sampled_snap_ids[1]) AND id >= $start_db_id;
    """
    result = execute(conn, query_nstars)
    n_stars = first(result).count
    println("Number of stars per snapshot: $n_stars")
    
    times = Float64[]
    virial_ratios = Float64[]
    virial_ratios_all = Float64[]
    bound_fractions = Float64[]
    
    for (idx, snap_id) in enumerate(sampled_snap_ids)
        print("\rAnalyzing snapshot $idx/$(length(sampled_snap_ids)) (ID: $snap_id)...")
        
        query_snap = """
            SELECT position, velocity, mass, time_myr FROM (
                SELECT id, position, velocity, mass, time_myr 
                FROM star_snapshots 
                WHERE dim_space = $dim_space AND snapshot_id = $snap_id AND id >= $start_db_id
                ORDER BY id DESC 
                LIMIT $n_stars
            ) as sub 
            ORDER BY id ASC;
        """
        result = execute(conn, query_snap)
        df = DataFrame(result)
        
        if nrow(df) == 0
            continue
        end
        
        pos = Array{Float64, 2}(undef, nrow(df), dim_space)
        vel = Array{Float64, 2}(undef, nrow(df), dim_space)
        mass = zeros(Float64, nrow(df))
        
        for i in 1:nrow(df)
            for d in 1:dim_space
                # Handle cases where LibPQ arrays might be nested or direct
                pos[i, d] = df.position[i][d]
                vel[i, d] = df.velocity[i][d]
            end
            mass[i] = df.mass[i]
        end
        
        mass ./= sum(mass)
        time_val = Float64(df.time_myr[1])
        
        bound_indices, T_bound, V_bound = get_bound_subsystem(pos, vel, mass, args["g-constant"], args["softening"])
        
        Q_bound = V_bound != 0 ? T_bound / abs(V_bound) : NaN
        
        # Entire cluster
        cm_pos_all = sum(pos .* mass, dims=1) ./ sum(mass)
        cm_vel_all = sum(vel .* mass, dims=1) ./ sum(mass)
        
        rel_pos_all = pos .- cm_pos_all
        rel_vel_all = vel .- cm_vel_all
        
        T_all = 0.5 * sum(mass .* vec(sum(rel_vel_all.^2, dims=2)))
        
        V_i_all = zeros(nrow(df))
        Threads.@threads for i in 1:nrow(df)
            phi_i = 0.0
            for j in 1:nrow(df)
                if i != j
                    dist_sq = 0.0
                    for d in 1:dim_space
                        dist_sq += (rel_pos_all[i, d] - rel_pos_all[j, d])^2
                    end
                    phi_i += mass[j] / sqrt(dist_sq + args["softening"]^2)
                end
            end
            V_i_all[i] = -args["g-constant"] * mass[i] * phi_i
        end
        
        V_all = 0.5 * sum(V_i_all)
        Q_all = V_all != 0 ? T_all / abs(V_all) : NaN
        
        bound_frac = length(bound_indices) / nrow(df)
        
        push!(times, time_val)
        push!(virial_ratios, Q_bound)
        push!(virial_ratios_all, Q_all)
        push!(bound_fractions, bound_frac)
    end
    
    println("\nAnalysis complete. Generating plots...")
    close(conn)
    
    p1 = plot(times, virial_ratios, color=:crimson, label="Bound Core Virial Q = T/|V|", linewidth=2)
    plot!(p1, times, virial_ratios_all, color=:darkorange, linestyle=:dash, label="Entire Cluster Virial Q = T/|V|", linewidth=2)
    hline!(p1, [0.5], color=:gray, linestyle=:dot, label="Virial Equilibrium (Q=0.5)")
    ylabel!(p1, "Virial Ratio Q")
    title!(p1, "Evolution of the Virial Ratio (N=$dim_space dimensions)")
    
    p2 = plot(times, bound_fractions .* 100, color=:royalblue, label="Bound Star %", linewidth=2)
    xlabel!(p2, "Time (Myr)")
    ylabel!(p2, "Bound Stars (%)")
    
    p = plot(p1, p2, layout=(2, 1), size=(1000, 800), link=:x)
    
    plot_filename = "virial_evolution_$(dim_space)D_julia.png"
    savefig(p, plot_filename)
    println("Plot successfully saved to '$plot_filename'!")
end

main()
